src/drivers/imu/invensense/icm20608g/ICM20608G.cpp

```diff
@@ -354,12 +370,13 @@ int ICM20608G::DataReadyInterruptCallback(int irq, void *context, void *arg)
 
 void ICM20608G::DataReady()
 {
-	perf_count(_drdy_interval_perf);
+	const uint8_t count = _drdy_count.fetch_add(1) + 1;
 
-	if (_data_ready_count.fetch_add(1) >= (_fifo_gyro_samples - 1)) {
-		_data_ready_count.store(0);
-		_fifo_watermark_interrupt_timestamp = hrt_absolute_time();
-		_fifo_read_samples.store(_fifo_gyro_samples);
+	uint8_t expected = 0;
+
+	// at least the required number of samples in the FIFO
+	if ((count >= _fifo_gyro_samples) && _drdy_fifo_read_samples.compare_exchange(&expected, _fifo_gyro_samples)) {
```

Comment:
I'd avoid `_drdy_fifo_read_samples` altogether and add the perf counter here directly. Or is there a specific reason you did it that way?

Reply:
We can enable lazy FPU stacking if floating point is avoided in interrupts. 
 - http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dai0298a/DAFGGBJD.html
 - https://github.com/PX4/Firmware/pull/14454